FuelLabs/fuel-indexer
# enhancement: add playground constraints

### Description
- PR adds:
  - A rate limit config to the indexer service
    - This is configurable
  - Rust doc comments
  - General cleanup
  - Adds timeout to indexer `handle_events` call
    - This is configurable

### Testing steps
⚠️ These testing steps test both  rate-limiting and handler execution timeouts

- [ ] Update your `fuel-indexer-test` indexer lib module to something that will DOS the service
- [ ] Ensure you rebuild the indexer `bash scripts/utils/build_test_wasm_module.bash`

```rust
extern crate alloc;

use fuel_indexer_macros::indexer;
use fuel_indexer_plugin::prelude::*;

#[indexer(
    manifest = "packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml"
)]
mod fuel_indexer_test {

    fn fuel_indexer_test_blocks(block_data: BlockData) {
        loop {}
    }
}
```

- [ ] Run a local fuel node `cargo run --bin fuel-node`
- [ ] Start the service connected to beta-3

```bash
cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```
  - ⚠️ On master using this ☝🏽indexer DOSes the service (which you should verify)
    - I use `htop` 
  - However, on this branch, you will see that the blocking indexer does not block the service 

#### Testing rate limiting
- [ ] Start the service with rate limiting enabled connected to beta-3
```bash
cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --run-migrations --rate-limit --rate-limit-rps 3 --rate-limit-window 10  --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```
- [ ] Create a little bash script to simulate someone blitzing the GraphQL API

```bash
#!/bin/bash

while True; do
    curl -X POST \
        -H "Content-Type: application/graphql" \
        --data '{ "query":"query { tx { id, block } }" }' \
        http://0.0.0.0:29987/api/graph/fuel_indexer_test/index1
done
```
- [ ] Call the bash script `bash my_script.bash`
  - You should notice that your requests get throttled  
  - You're not gonna get an error response status because the `BufferLayer` is not rejecting requests, rather it's queue'ing your requests 
  - ⚠️ On master using this ☝🏽 script just blitzes the web server (which you should verify)

### Changelog
- enhancement: add playground constraints
